cli/command/system/prune_progress.go

```diff
@@ -0,0 +1,175 @@
+				ID:         "Progress",
+				Action:     s,
+				Current:    i,
+				Total:      int64(numEvents),
+				HideCounts: true,
+			}
+			if err := progressOut.WriteProgress(prog); err != nil {
+				progressStruct.errorChan <- err
+				return
+			}
+			i++
+		}
+	}()
+
+	if err := jsonmessage.DisplayJSONMessagesToStream(pipeReader, dockerCli.Out(), nil); err != nil {
+		progressStruct.errorChan <- err
+		return
+	}
+}
+
+func countObjectsToRemove(dockerCli command.Cli, options pruneOptions) (int, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	volCount := 0
+	if options.pruneVolumes {
+		//Volumes listing call
+		volFilters := filters.NewArgs()
+		vols, err := dockerCli.Client().VolumeList(ctx, volFilters)
+		if err != nil {
+			return 0, err
+		}
+		volCount = len(vols.Volumes)
+	}
+
+	//Containers listing call.
+	//We get all the containers because we need to check which images are attached to them.
+	cntFilters := filters.NewArgs()
+	cntOpts := types.ContainerListOptions{
+		Quiet:   true,
+		Size:    false,
+		All:     true,
+		Latest:  false,
+		Since:   "",
+		Before:  "",
+		Limit:   -1,
+		Filters: cntFilters,
+	}
+	cnts, err := dockerCli.Client().ContainerList(ctx, cntOpts)
+	if err != nil {
+		return 0, err
+	}
+
+	cntCount := 0
+	//We check which containers are unused to know which images are unused, also.
+	imgUsage := make(map[string]int)
+	for _, cnt := range cnts {
+		if cnt.State == "running" {
+			imgUsage[cnt.ImageID]++
+		} else {
+			imgUsage[cnt.ImageID] += 0
+			cntCount++
+		}
+	}
+
+	//Images listing call.
+	imgFilters := filters.NewArgs()
+	if !options.all {
+		imgFilters.Add("dangling", "true")
+	}
+	imgOpts := types.ImageListOptions{
+		All:     options.all,
+		Filters: imgFilters,
+	}
+	imgs, err := dockerCli.Client().ImageList(ctx, imgOpts)
+	if err != nil {
+		return 0, err
+	}
+
+	imgCount := 0
+	//Here we merge the images from the containers call with the images listing call.
+	for _, img := range imgs {
+		v, ok := imgUsage[img.ID]
+		if !ok || v == 0 {
+			imgCount++
+		}
+	}
+
+	//Network listing call.
+	ntwFilters := filters.NewArgs()
+	ntwFilters.Add("type", "custom")
+	ntwOpts := types.NetworkListOptions{
+		Filters: ntwFilters,
+	}
+	ntws, err := dockerCli.Client().NetworkList(ctx, ntwOpts)
+	if err != nil {
+		return 0, err
+	}
+
+	return cntCount + volCount + imgCount + len(ntws), nil
+}
+
+func monitorRemovingEvents(dockerCli command.Cli, out chan<- string, progressChans *pruneProgressChan) {
+	fArgs := filters.NewArgs(
+		filters.Arg("type", "container"),
+		filters.Arg("type", "volume"),
+		filters.Arg("type", "network"),
+		filters.Arg("type", "image"),
+		//Destroy for container, network and volume.
+		filters.Arg("event", "destroy"),
+		//Delete only for image.
+		filters.Arg("event", "delete"),
+	)
+	eOpts := types.EventsOptions{
+		Since:   "",
+		Until:   "",
+		Filters: fArgs,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
This function should accept a `WithCancel` `ctx` as an argument so that the context can be shared with `runAllPrune` (also needs to be fixed in another PR). 

I think the `defer cancel()` should probably be in the `runPrune()` so that this goroutine and the output goroutines will exit if someone `ctrl-c` the main thread. I'm not sure that is working correctly right now, I'll need to try it out.